src/rpc/blockchain.h

```diff
@@ -47,11 +50,7 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
 /** Used by getblockstats to get feerates at different percentiles by weight  */
 void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES], std::vector<std::pair<CAmount, int64_t>>& scores, int64_t total_weight);
 
-//! Pointer to node state that needs to be declared as a global to be accessible
-//! RPC methods. Due to limitations of the RPC framework, there's currently no
-//! direct way to pass in state to RPC methods without globals.
-extern NodeContext* g_rpc_node;
-
-CTxMemPool& EnsureMemPool();
+NodeContext& EnsureNodeContext(const util::Ref& context);
```

Comment:
Have you considered `EnsureNodeContext(const JSONRPCRequest& req)` instead? Same for `EnsureMemPool`. Could avoid forward declaration above.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18740#discussion_r413428813

> Have you considered `EnsureNodeContext(const JSONRPCRequest& req)` instead? Same for `EnsureMemPool`. Could avoid forward declaration above.

Probably considered it, but I don't like this style (along lines of c++ koan https://www.youtube.com/watch?v=Zx_Tjp9WIII&t=15m07s)